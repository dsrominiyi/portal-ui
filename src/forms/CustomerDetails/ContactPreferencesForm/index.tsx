import React, { FunctionComponent, useState, useEffect } from 'react';

import QuestionBox from '../../../components/QuestionBox';
import MultiChoice, { updateSelectedOptions } from '../../../components/MultiChoice';
import Dropdown from '../../../components/Dropdown';

import {
  consentTypeOptions,
  contactMethodOptions,
  referredByOptions
} from '../../../helpers/optionLists';

import formStyles from '../../styles.scss';

interface Props {
  contactDetails: Contact[];
  onChange(state: ContactPreferences): void;
}

type UseStateArray = [ContactPreferences, React.Dispatch<React.SetStateAction<ContactPreferences>>];

export const ContactPreferencesForm: FunctionComponent<Props> = ({
  onChange,
  contactDetails
}: Props) => {
  const [state, setState]: UseStateArray = useState({
    marketingConsentType: consentTypeOptions[0],
    marketingContactMethods: [] as ContactMethod[],
    nonMarketingContactMethods: [] as ContactMethod[],
    referredBy: ''
  });

  useEffect(() => {
    onChange(state);
  }, [state.marketingContactMethods, state.nonMarketingContactMethods, state.referredBy]);

  const update = (field: string, value: string | string[]) => {
    setState({ ...state, [field]: value });
  };

  const createDisabledOptionsValue = (): Array<string> => {
    const disabledContacts = new Set([
      ContactMethod.Email,
      ContactMethod.Phone,
      ContactMethod.TextMessage
    ]);

    contactDetails.forEach(contact => {
      if (
        [ContactType.HomeNumber, ContactType.WorkNumber, ContactType.MobileNumber].includes(
          contact.type
        )
      ) {
        if (ContactType.MobileNumber === contact.type) {
          disabledContacts.delete(ContactMethod.TextMessage);
        }
        disabledContacts.delete(ContactMethod.Phone);
      }

      if ([ContactType.WorkEmail, ContactType.PersonalEmail].includes(contact.type)) {
        disabledContacts.delete(ContactMethod.Email);
      }
    });

    return Array.from(disabledContacts);
  };

  useEffect(() => {
    setState({
      ...state,
      marketingContactMethods: contactMethodOptions.filter(
        item => !createDisabledOptionsValue().includes(item)
      ),
      nonMarketingContactMethods: contactMethodOptions.filter(
        item => !createDisabledOptionsValue().includes(item)
      )
    });
    onChange(state);
  }, [state.marketingConsentType, contactDetails]);

  const checkContactMethod = (): boolean => {
    if (
      state.marketingConsentType === MarketingConsentType.Unsubscribe ||
      state.marketingConsentType === MarketingConsentType.LegitimateInterest
    ) {
      if (state.marketingContactMethods.length !== 0) update('marketingContactMethods', []);

      return false;
    }
    return true;
  };

  return (
    <div className={formStyles.formSection}>
      <h3>Marketing/Contact Preferences</h3>
      <QuestionBox
        text="Marketing preference consent type"
        inputFields={[
          <MultiChoice
            key="0"
            options={consentTypeOptions}
            selectedOption={state.marketingConsentType}
            onClick={option => update('marketingConsentType', option)}
            disabledOptions={
              state.marketingConsentType !== consentTypeOptions[0] ? [consentTypeOptions[0]] : []
            }
            attributes={{ 'test-id': 'select-marketing-consent-type' }}
          />
        ]}
      />
      {checkContactMethod() && (
        <QuestionBox
          text="Marketing contact methods"
          inputFields={[
            <MultiChoice
              key="0"
              multiSelect
              options={contactMethodOptions}
              disabledOptions={createDisabledOptionsValue()}
              selectedOptions={state.marketingContactMethods}
              onClick={option => {
                update(
                  'marketingContactMethods',
                  updateSelectedOptions(state.marketingContactMethods, option)
                );
              }}
              attributes={{ 'test-id': 'select-marketing-contact-method' }}
            />
          ]}
        />
      )}
      <QuestionBox
        text="Non marketing contact methods"
        inputFields={[
          <MultiChoice
            key="0"
            multiSelect
            options={contactMethodOptions}
            disabledOptions={createDisabledOptionsValue()}
            selectedOptions={state.nonMarketingContactMethods}
            onClick={option => {
              update(
                'nonMarketingContactMethods',
                updateSelectedOptions(state.nonMarketingContactMethods, option)
              );
            }}
            attributes={{ 'test-id': 'select-non-marketing-contact-method' }}
          />
        ]}
      />
      <QuestionBox
        inputFields={[
          <Dropdown
            key="0"
            label="Where did you hear about us?"
            options={referredByOptions}
            selected={state.referredBy}
            onChange={value => update('referredBy', String(value))}
            attributes={{ 'test-id': 'select-referred-by' }}
          />
        ]}
      />
    </div>
  );
};

ContactPreferencesForm.displayName = 'ContactPreferencesForm';
export default ContactPreferencesForm;
