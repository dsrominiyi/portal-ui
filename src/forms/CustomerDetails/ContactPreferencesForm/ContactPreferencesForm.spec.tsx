import React from 'react';
import { shallow, mount, ShallowWrapper, ReactWrapper } from 'enzyme';
import { act } from 'react-dom/test-utils';

import ContactPreferencesForm from '.';

let form: ShallowWrapper | ReactWrapper;
const onChange = jest.fn();

const getField = (fieldIndex: number) => {
  return (form
    .find('QuestionBox')
    .at(fieldIndex)
    .props() as any).inputFields[0];
};

describe('ContactPreferencesForm', () => {
  const contactDetails = [
    { type: ContactType.HomeNumber, detail: '02079460111' },
    { type: ContactType.PersonalEmail, detail: 'test@example.com' },
    { type: ContactType.MobileNumber, detail: '02079460111' }
  ];
  beforeEach(() => {
    jest.clearAllMocks();
    form = shallow(<ContactPreferencesForm onChange={onChange} contactDetails={contactDetails} />);
  });

  test('ContactPreferencesForm matches snapshot', () => {
    expect(form).toMatchSnapshot();
  });

  test('selecting a marketing consent type updates the marketingConsentType state value', () => {
    const option = 'Email';
    let consentMultiChoice = getField(0);
    consentMultiChoice.props.onClick(option);
    consentMultiChoice = getField(0);

    expect(consentMultiChoice.props.selectedOption).toBe(option);
  });

  test('selecting a marketing contact method updates the marketingContactMethods state value', () => {
    const option = 'Email';
    const option2 = 'Phone';
    let marketingContactMultiChoice = getField(1);
    marketingContactMultiChoice.props.onClick(option);
    marketingContactMultiChoice = getField(1);

    expect(marketingContactMultiChoice.props.selectedOptions).toEqual([option]);

    marketingContactMultiChoice.props.onClick(option2);
    marketingContactMultiChoice = getField(1);

    expect(marketingContactMultiChoice.props.selectedOptions).toEqual([option, option2]);

    marketingContactMultiChoice.props.onClick(option);
    marketingContactMultiChoice = getField(1);

    expect(marketingContactMultiChoice.props.selectedOptions).toEqual([option2]);
  });

  test('selecting a non-marketing contact method updates the noMarketingContactMethods state value', () => {
    const option = 'Email';
    const option2 = 'Phone';
    const marketingPreferenceConsentType = getField(0);
    marketingPreferenceConsentType.props.onClick(option2);

    let nonMarketingContactMultiChoice = getField(1);
    nonMarketingContactMultiChoice.props.onClick(option);
    nonMarketingContactMultiChoice = getField(1);

    expect(nonMarketingContactMultiChoice.props.selectedOptions).toEqual([option]);

    nonMarketingContactMultiChoice.props.onClick(option2);
    nonMarketingContactMultiChoice = getField(1);

    expect(nonMarketingContactMultiChoice.props.selectedOptions).toEqual([option, option2]);

    nonMarketingContactMultiChoice.props.onClick(option);
    nonMarketingContactMultiChoice = getField(1);

    expect(nonMarketingContactMultiChoice.props.selectedOptions).toEqual([option2]);
  });

  test('selecting a referred by option updates the referredBy state value', () => {
    const value = 'Recommended';
    let referredByDropdown = getField(2);

    referredByDropdown.props.onChange(value);
    referredByDropdown = getField(2);

    expect(referredByDropdown.props.selected).toBe(value);
  });

  test('onChange is called with the form state when any field is updated', () => {
    form = mount(<ContactPreferencesForm onChange={onChange} contactDetails={contactDetails} />);

    const marketingConsentType = ContactMethod.Email;
    const consentMultiChoice = getField(0);

    act(() => {
      consentMultiChoice.props.onClick(marketingConsentType);
    });
    form.update();

    expect(onChange).toHaveBeenCalledWith({
      marketingConsentType,
      marketingContactMethods: [
        ContactMethod.TextMessage,
        ContactMethod.Phone,
        ContactMethod.Email,
        ContactMethod.Post
      ],
      nonMarketingContactMethods: [
        ContactMethod.TextMessage,
        ContactMethod.Phone,
        ContactMethod.Email,
        ContactMethod.Post
      ],
      referredBy: ''
    });

    form.update();
    const marketingContactMethod = ContactMethod.Post;
    const marketingContactMultiChoice = getField(1);

    act(() => {
      marketingContactMultiChoice.props.onClick(marketingContactMethod);
    });

    expect(onChange).toHaveBeenCalledWith({
      marketingConsentType,
      marketingContactMethods: [
        ContactMethod.TextMessage,
        ContactMethod.Phone,
        ContactMethod.Email
      ],
      nonMarketingContactMethods: [
        ContactMethod.TextMessage,
        ContactMethod.Phone,
        ContactMethod.Email,
        ContactMethod.Post
      ],
      referredBy: ''
    });

    form.update();
    const nonMarketingContactMethod = ContactMethod.TextMessage;
    const nonMarketingContactMultiChoice = getField(2);

    act(() => {
      nonMarketingContactMultiChoice.props.onClick(nonMarketingContactMethod);
    });

    expect(onChange).toHaveBeenCalledWith({
      marketingConsentType,
      marketingContactMethods: [
        ContactMethod.TextMessage,
        ContactMethod.Phone,
        ContactMethod.Email
      ],
      nonMarketingContactMethods: [ContactMethod.Phone, ContactMethod.Email, ContactMethod.Post],
      referredBy: ''
    });

    form.update();
    const referredBy = 'Recommended';
    const referredByDropdown = getField(3);

    act(() => {
      referredByDropdown.props.onChange(referredBy);
    });

    expect(onChange).toHaveBeenCalledWith({
      marketingConsentType,
      marketingContactMethods: [
        ContactMethod.TextMessage,
        ContactMethod.Phone,
        ContactMethod.Email
      ],
      nonMarketingContactMethods: [ContactMethod.Phone, ContactMethod.Email, ContactMethod.Post],
      referredBy
    });
  });

  test('Check "Marketing contact methods" QuestionBox disabled status when contact details is empty', () => {
    form = mount(<ContactPreferencesForm onChange={onChange} contactDetails={[]} />);
    const marketingConsentType = getField(0);

    act(() => {
      marketingConsentType.props.onClick(MarketingConsentType.Phone);
    });
    form.update();

    const marketingContactMethods = getField(1);

    expect(marketingContactMethods.props.disabledOptions).toContain(ContactMethod.Phone);
    expect(marketingContactMethods.props.disabledOptions).toContain(ContactMethod.Email);
    expect(marketingContactMethods.props.disabledOptions).toContain(ContactMethod.TextMessage);
  });

  test('Check "Marketing contact methods" QuestionBox disabled status when contact details has a mobile number', () => {
    form = mount(
      <ContactPreferencesForm onChange={onChange} contactDetails={[contactDetails[2]]} />
    );
    const marketingConsentType = getField(0);

    act(() => {
      marketingConsentType.props.onClick(MarketingConsentType.Phone);
    });
    form.update();

    const marketingContactMethods = getField(1);

    expect(marketingContactMethods.props.disabledOptions).toContainEqual(ContactMethod.Email);
  });

  test('Check "Marketing contact methods" QuestionBox disabled status when contact details has a email address and home phone number', () => {
    const contactDetailsWithPhoneAddress = [contactDetails[0], contactDetails[1]];
    form = mount(
      <ContactPreferencesForm onChange={onChange} contactDetails={contactDetailsWithPhoneAddress} />
    );
    const marketingConsentType = getField(0);

    act(() => {
      marketingConsentType.props.onClick(MarketingConsentType.Phone);
    });
    form.update();

    const marketingContactMethods = getField(1);

    expect(marketingContactMethods.props.disabledOptions).toContain(ContactMethod.TextMessage);
  });

  test('Reset marketingContactMethods state when user select Unsubscribe option (hide the Marketing contact methods)', () => {
    const contactDetailsWithPhoneAddress = [contactDetails[0], contactDetails[1]];
    form = mount(
      <ContactPreferencesForm onChange={onChange} contactDetails={contactDetailsWithPhoneAddress} />
    );
    const marketingConsentType = getField(0);

    act(() => {
      marketingConsentType.props.onClick(MarketingConsentType.Phone);
    });
    form.update();

    const marketingContactMethods = getField(1);

    act(() => {
      marketingContactMethods.props.onClick(ContactMethod.Phone);
      marketingConsentType.props.onClick(MarketingConsentType.Unsubscribe);
    });
    form.update();

    const nonMarketingContactMethods = getField(1);

    expect(nonMarketingContactMethods.props.selectedOptions.length).toEqual(3);
  });

  test('Hide Marketing preferences when Unsubscribe, or Legitimate Interest option selected', () => {
    form = mount(<ContactPreferencesForm onChange={onChange} contactDetails={[]} />);
    const marketingConsentType = getField(0);
    const marketingContactMethodsTestId = { 'test-id': 'select-marketing-contact-method' };

    expect(marketingConsentType.props.selectedOption).toBe(MarketingConsentType.LegitimateInterest);
    expect(form.find(marketingContactMethodsTestId).length).toEqual(0);

    act(() => {
      marketingConsentType.props.onClick(MarketingConsentType.Unsubscribe);
    });
    form.update();

    expect(form.find(marketingContactMethodsTestId).length).toEqual(0);

    act(() => {
      marketingConsentType.props.onClick(MarketingConsentType.Email);
    });
    form.update();
    expect(form.find(marketingContactMethodsTestId).length).toEqual(1);
  });
});
