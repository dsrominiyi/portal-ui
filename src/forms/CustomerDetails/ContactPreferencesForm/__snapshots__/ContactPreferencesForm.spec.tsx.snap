// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactPreferencesForm ContactPreferencesForm matches snapshot 1`] = `
<div
  className="formSection"
>
  <h3>
    Marketing/Contact Preferences
  </h3>
  <QuestionBox
    inputFields={
      Array [
        <MultiChoice
          attributes={
            Object {
              "test-id": "select-marketing-consent-type",
            }
          }
          disabledOptions={Array []}
          onClick={[Function]}
          options={
            Array [
              "Legitimate Interest",
              "Face to Face",
              "Phone",
              "Email",
              "Unsubscribe",
            ]
          }
          selectedOption="Legitimate Interest"
        />,
      ]
    }
    text="Marketing preference consent type"
  />
  <QuestionBox
    inputFields={
      Array [
        <MultiChoice
          attributes={
            Object {
              "test-id": "select-non-marketing-contact-method",
            }
          }
          disabledOptions={Array []}
          multiSelect={true}
          onClick={[Function]}
          options={
            Array [
              "Text Message",
              "Phone",
              "Email",
              "Post",
            ]
          }
          selectedOptions={Array []}
        />,
      ]
    }
    text="Non marketing contact methods"
  />
  <QuestionBox
    inputFields={
      Array [
        <Dropdown
          attributes={
            Object {
              "test-id": "select-referred-by",
            }
          }
          label="Where did you hear about us?"
          onChange={[Function]}
          options={
            Array [
              "Practical Caravan",
              "Lead Linkedin",
              "C.V.R",
              "Autotrader Motorhome",
              "Yellow Pages",
              "Cross-selling",
              "Existing Business",
              "Lapsed Caravan",
              "Lapse CV",
              "Lapsed Home",
              "Ramsey Cricket Club",
              "Central Marketing",
              "Come Back Letter",
              "Fleet",
              "Stuarts Mailings",
              "Truck Tele",
              "Passing Trade",
              "Old Magazine Adverts",
              "Coachman Club",
              "Referral From NHF",
              "Recommended",
              "Vernon Property",
              "Insurecom Broker",
            ]
          }
          selected=""
        />,
      ]
    }
  />
</div>
`;
